service/src/test/java/bio/terra/workspace/db/FolderDaoTest.java

```diff
@@ -0,0 +1,347 @@
+package bio.terra.workspace.db;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import bio.terra.common.exception.BadRequestException;
+import bio.terra.common.exception.MissingRequiredFieldException;
+import bio.terra.workspace.common.BaseUnitTest;
+import bio.terra.workspace.db.exception.DuplicateFolderDisplayNameException;
+import bio.terra.workspace.db.exception.DuplicateFolderIdException;
+import bio.terra.workspace.db.exception.FolderNotFoundException;
+import bio.terra.workspace.db.exception.WorkspaceNotFoundException;
+import bio.terra.workspace.service.folder.model.Folder;
+import bio.terra.workspace.service.workspace.model.Workspace;
+import bio.terra.workspace.service.workspace.model.WorkspaceStage;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class FolderDaoTest extends BaseUnitTest {
```

Comment:
I didn't see this file yesterday. I see you have a bunch of tests here. FYI for major functionality (create folder fails when workspace doesn't exist) I will ask for controller test. That way, it's tested regardless of whether logic lives in controller, service, or dao. I don't have to know how a particular endpoint is implemented, I'll just always look in controller test. For me, tests don't also need to be here. Ideally, we would not duplicate tests.

Reply:
Unit test should test what that corresponding class is doing. 

the controller does auth checking and convert folder to ApiFolder so that should be the thing to test in controller test. 

The implementation details like duplicate display name handling, folder not found,  cycles etc are folderDao implementation details so i prefer them to be in the folderDao. If we want to dedup, I'd prefer to remove them from the controllerTest

Reply:
To me, this is another rule that's not hard-and-fast.

I think of our controller tests as testing a black box. We call createFolder like a user would call it; we return what user would get. Those are the most important tests. A Dao unit test could pass, but the controller test could fail, if there's a bug in controller or service layer.

For something like cycle, sure -- that's a corner case, that could be in Dao test.

It's strange to me if "Create X fails if workspace doesn't exist" test is in controller test, service test, or dao test, depending on where code is implemented.

Interesting discussion. :) (I'm not requiring any changes.)

Reply:
From https://g3doc.corp.google.com/cloud/control/g3doc/users/style/unit_tests.md#prefer-fakes-over-mocks

black box test which tests the public interface and behaviors

Black box test is almost always preferable to white box test,

Reply:
i don't see how the discussion for fakes vs mocks applies here.

the services are still mocked out in the folderApiControllerTest. The controller shouldn't worry about the implementation details being brokens in the folderDao. Say there are three controllers calling folderDao. I don't think all three should need to / able to cover all the edge cases as they are guaranteed through the unit test.

But i indeed caught bugs in the controller test for the controller code.

Reply:
I see your point, I guess I have different paranoia level. :)

In FolderApiControllerTest, `SamService` is mocked because it talks to SAM. It uses real Folder controller, Folder service, and Folder dao. So it is mimicing what user does.

If my Controller implements CreateFoo endpoint, Controller test should have a test for createFoo(). I think we're in agreement there.

If CreateFoo throws 403 for certain case, I would add controller test for that.

If a second Controller implements CreateBar, and also uses folderDao, I would add a throws404 controller test for that second API. I see your point that it's duplicated, so fine not to add. But I would add because:

- Technically there could be a bug with CreateBar throws 404, that is not caught by Dao test. In my experience, when you have tests for bits and pieces of your stack, but you don't test the whole stack, you miss bugs. The folder code is simple, but you can imagine for other endpoints, there's complex interaction between various layers.
- To me, it's nice if there are controller tests for everything that's advertised in yaml

Reply:
i agree 403 needs to be catch at controller level, added 403 tests for all the controller test.

i think we are in general agreement that test coverage should be thorough across the stack. I like the codedacity scanner that some broad repo is using.